### ISSUE TITLE
add "Obtain backtrace of complex error scenarios" example

### ISSUE THREAD

**Issue by budziq** [rust-lang-nursery/rust-cookbook#216]
Show some realistic 3 level error handling (possibly based on largely stripped down version of [Download a file example](https://brson.github.io/rust-cookbook/net.html#download-a-file-to-a-temporary-directory) without tempdir/headder noise)

Use the error-chain crate

cc https://github.com/brson/rust-cookbook/issues/184

**Comment by ludwigpacifici** [comment]
Hello @budziq 

I would be interested to provide an example for this topic.

**Comment by budziq** [comment]
Thanks @ludwigpacifici . the link was for the previous landing page sry.

**Comment by ludwigpacifici** [comment]
To make sure I am heading in the right direction, can you tell me if the snippet highlights a backtrace with 3 levels of error handling?

```rust
fn run() -> Result<()> {
    f1().chain_err(|| "run failed beacuse f1")?;
    Ok(())
}

fn f1() -> Result<std::fs::File> {
    f2().chain_err(|| "f1 failed because of f2")
}

fn f2() -> Result<std::fs::File> {
    File::open("oops").chain_err(|| "f2 is unable to open 'oops' file")
}
```

Ouput:

```text
Error: run failed beacuse f1
Caused by: f1 failed because of f2
Caused by: f2 is unable to open 'oops' file
Caused by: No such file or directory (os error 2)
```

**Comment by budziq** [comment]
@ludwigpacifici pretty much yeah but I would like for this example to have very simple but at least semi realistic case and a little nicer function names. Would you be able to come up with and idea for it? I'm thinking of some high level operation failing because obtaining a token (parsing possibly an int or just utf8 string) from file failed due to open error.

**Comment by ludwigpacifici** [comment]
Write an implementation of [ROT13](https://en.wikipedia.org/wiki/ROT13). It takes a file as input, and print the encoded text on the standard output. To keep the code short, it will switch [a-zA-Z] and other valid characters are skipped. At first glance, it can fails if the file cannot be opened or a string from the file is not utf8.

**Comment by budziq** [comment]
I would suggest against ROT13 I would hesitate advertising this manner of ~~encryption~~/obfuscation even in joking manner. But any other failable mode of text processing would do such as matching file contents against a regexp?

**Comment by ludwigpacifici** [comment]
Then, what do you think of a basic calculator: `<number><op><number>`, where `<number>` is a decimal integer and `<op>` is one of `+`, `-`, `/` and `*`?

```
$ ./calculator 2+3
=> 5
$ ./calculator 2+b3
=> backstrace
```

**Comment by budziq** [comment]
ok lets do it and we'll see how it goes.
But please remember that the example is focused on error handling (3 levels at least). and much else will be noise. we might have to do similarly as in the case of [fractal drawing example](https://rust-lang-nursery.github.io/rust-cookbook/concurrency.html#ex-threadpool-fractal) where the actual drawing is hidden in order not to obfuscate the threadpool example

**Comment by ludwigpacifici** [comment]
Hello @budziq,

About the calculator, I can have just one level of error, which is not great for this issue. I'll think about other examples.

**Comment by budziq** [comment]
@ludwigpacifici don't get discouraged!  If the calculator is not a perfect example, maybe you could think about some other super simple usecase?

**Comment by ludwigpacifici** [comment]
#228 could be a good candidate :-) Deserialize a csv integer into an incorrect type.

Csv data looks like:

```csv
year,...
2017,...
```

And a faulty Rust struct:

```rust
#[derive(Deserialize, ...)]
struct Record {
    year: u8, // u8 is too small to contain a year.
    //...
}
```

The backtrace should be good enough for this issue:

```text
Error: Could not read csv data
Caused by: Cannot deserialize record
Caused by: CSV deserialize error: record 1 (line: 2, byte: 28): field 0: number too large to fit in target type
Caused by: field 0: number too large to fit in target type
```

Let me know what you think.

**Comment by budziq** [comment]
looks very cool :+1:

**Comment by ludwigpacifici** [comment]
Can be close. It will make the always expanding issue list slightly smaller :-)

**Comment by budziq** [comment]
Right thanks! Please add "fixes #216" to the PR's in the future so the connected issue will autoclose :)
